### PR TITLE
feat: Update collection arguments

### PR DESCRIPTION
**Pull request by sean-m-sullivan**
##### SUMMARY
Update the injectors for the AAP credential type to work across collections

##### ISSUE TYPE
- New or Enhanced Feature

##### COMPONENT NAME
 - Collection

##### AWX VERSION

```
24.6.1
```

##### ADDITIONAL INFORMATION
Related PR's:

- [awx-plugins](https://github.com/ansible/awx-plugins/pull/55)
- [ansible.platform](https://github.com/ansible/aap-gateway/pull/627)
- [ansible.hub](https://github.com/ansible-collections/ansible_hub/pull/29)
- [ansible.controller](https://github.com/ansible/awx/pull/15627)
- [ansible.eda](https://github.com/ansible/event-driven-ansible/pull/347)
- [eda-server](https://github.com/ansible/eda-server/pull/1162)
- infra.aap_configuration - Done.
- https://issues.redhat.com/browse/ANSTRAT-1181

### CODE REVIEW

**Comment by AlanCoding** [comment]
Consistency is the most important thing for this. What other content is using these connection parameters? This?

https://galaxy.ansible.com/ui/repo/published/infra/aap_configuration/docs/

**Comment by sean-m-sullivan** [comment]
> Consistency is the most important thing for this. What other content is using these connection parameters? This?
> 
> https://galaxy.ansible.com/ui/repo/published/infra/aap_configuration/docs/

This is the primary thing using it at the moment, and the installer using it as well, it was more Some Variables should be exposed to playbooks for use, lets set a standard.

**Comment by AlanCoding** [comment]
Looking good :+1: 

I got some more background on this initiative. It looks like you've already taken the lead in proposing these changes elsewhere. I fully support this, but want to wait a few days to assure we've solidified consensus.

**Comment by AlanCoding** [comment]
I'm actually not convinced this will work anymore. I expect that if the collection receives `AAP_HOSTNAME` then in that scenario specifically, it needs to append to use `/api/controller/` as opposed to `/api/`. I don't see that.

I could test this manually to clarify specifically how it's not working, but there's no automated tests we can rely on right now.

**Comment by AlanCoding** [comment]
> https://github.com/ansible/event-driven-ansible/pull/347

The PR you created here is for the collection. There may be a needed change to eda-server as well.

https://github.com/ansible/eda-server/blob/ac4be90e413196e922bd255a04589711b25b818e/src/aap_eda/core/management/commands/create_initial_data.py#L302-L315

This is probably more consequential, since this is a credential type used in-app, most similar to the awx_plugins change you are making. At some point we hope to essentially merge the 2 CredentialType models, but we are not there yet, so the controller type is more-or-less duplicated in both places.

**Comment by sean-m-sullivan** [comment]
I tested all of these changes using my modifications, and the test plays,
they work, can go into more detail

**Comment by sean-m-sullivan** [comment]
It may be needed, didn't know it was there will look  and add that to the list

> > https://github.com/ansible/event-driven-ansible/pull/347
> 
> The PR you created here is for the collection. There may be a needed change to eda-server as well.
> 
> https://github.com/ansible/eda-server/blob/ac4be90e413196e922bd255a04589711b25b818e/src/aap_eda/core/management/commands/create_initial_data.py#L302-L315
> 
> This is probably more consequential, since this is a credential type used in-app, most similar to the awx_plugins change you are making. At some point we hope to essentially merge the 2 CredentialType models, but we are not there yet, so the controller type is more-or-less duplicated in both places.

File: awx_collection/plugins/doc_fragments/auth.py [hunk 1]
```diff
@@ -19,32 +19,43 @@ class ModuleDocFragment(object):
...
+  controller_oauthtoken:
+    description:
+    - The OAuth token to use.
+    - This value can be in one of two formats.
+    - A string which is the token itself. (i.e. bqV5txm97wqJqtkxlMkhQz0pKhRMMX)
+    - A dictionary structure as returned by the token module.
+    - If value not set, will try environment variable C(CONTROLLER_OAUTH_TOKEN) and then config files
+    type: raw
+    version_added: "3.7.0"
+    aliases: [ tower_oauthtoken, aap_token ]
```
Comment (AlanCoding) [inline awx_collection/plugins/doc_fragments/auth.py]: I would prefer for the main name to be `aap_token`, otherwise looks alright.

Comment (sean-m-sullivan) [reply awx_collection/plugins/doc_fragments/auth.py]: Updated

File: awx_collection/plugins/doc_fragments/auth.py [hunk 2]
```diff
@@ -19,32 +19,43 @@ class ModuleDocFragment(object):
...
+  aap_token:
+    description:
+    - The OAuth token to use.
+    - This value can be in one of two formats.
+    - A string which is the token itself. (i.e. bqV5txm97wqJqtkxlMkhQz0pKhRMMX)
+    - A dictionary structure as returned by the token module.
+    - If value not set, will try environment variable C(CONTROLLER_OAUTH_TOKEN) and then config files
+    type: raw
+    version_added: "3.7.0"
+    aliases: [ tower_oauthtoken, controller_oauthtoken]
```
Comment (PabloHiro) [inline awx_collection/plugins/doc_fragments/auth.py]: Does this still make sense?

File: awx_collection/plugins/doc_fragments/auth_plugin.py [hunk 3]
```diff
@@ -17,45 +17,51 @@ class ModuleDocFragment(object):
...
+  oauth_token:
+    description:
+    - The OAuth token to use.
+    env:
+    - name: CONTROLLER_OAUTH_TOKEN
+      deprecated:
+        collection_name: 'awx.awx'
+        version: '4.0.0'
+        why: Collection name change
+        alternatives: 'TOWER_OAUTH_TOKEN, AAP_TOKEN'
```
Comment (PabloHiro) [inline awx_collection/plugins/doc_fragments/auth_plugin.py]: Same with this. Not sure how to handle this so it is clear that OAUTH is coming from the Resource Server but is not available in Controller

File: awx_collection/plugins/doc_fragments/auth.py (hunk 2 above)
Comment (AlanCoding) [reply awx_collection/plugins/doc_fragments/auth.py]: agree, I think this is okay if it has no aliases, due to other recent changes

**Comment by AlanCoding** [comment]
Here is a patch to hopefully fix the sanity failures.

https://github.com/sean-m-sullivan/awx/pull/60

Checks do have trouble running on forks, so we may not get meaningful output from that until merged here. I guess I can test separately... https://github.com/ansible/awx/pull/15729

**Review by AlanCoding — CHANGES_REQUESTED** [review]
I need to formally document the situation with tokens here. These lines of code were _removed_ in recent changes:

https://github.com/ansible/awx/commit/3ba6e2e394be319d41c80243f88399dc2bf2e4a0#diff-87795928c0b98b89601c8cc0bb837dd70788e1a0c7c5803a6bf7a695bc02984dL505-L507

```python
        if self.oauth_token:
            # If we have a oauth token, we just use a bearer header
            headers['Authorization'] = 'Bearer {0}'.format(self.oauth_token)
```

Changing how we parse the token argument is ignoring the problem. With no code that constructs the headers to use a token, you can't make requests using a token.

For now, I would say kick the token out of scope of this, and an issue can be filed for that. Current reading is that all params related to OAuth2 tokens here are non-functional and that's the blocker. So the simple change I'm asking for is to remove the token params.

**Comment by AlanCoding** [comment]
For sanity test failures, I've now realized (after banging my head on my desk for a bit) that the following failure is introduced by _this_ PR.

```
ERROR: Found 2 pep8 issue(s) which need to be resolved:
ERROR: plugins/module_utils/controller_api.py:80:14: E131: continuation line unaligned for hanging indent
ERROR: plugins/module_utils/controller_api.py:81:14: E131: continuation line unaligned for hanging indent
```

The patch I linked (my PR in your fork) is still valid. This is yet another fix that's needed for sanity to pass.

**Comment by AlanCoding** [comment]
For yet another request here, it was agreed that we need _some_ measure of testing of the new params. This is my proposed solution. It can be expanded to some other params, but cannot be expanded realistically to test env vars (maybe? you could try)

https://github.com/sean-m-sullivan/awx/pull/61

I think that will pass, so this (along with my other PR 60) should hopefully be easy inclusions without breaking checks.

To recap the full work list here:

 - fork PR 60 - 1 of 2 sanity test fixes
 - 2 of 2 sanity test fixes (from pep8)
 - fork PR 61 - add some very basic coverage for the updated params (that values given take effect)
 - remove all the params about using tokens, because they don't work

And I hope all that should cover it to get this ready to merged and moved along.

File: awx_collection/plugins/doc_fragments/auth_plugin.py (hunk 3 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/doc_fragments/auth_plugin.py]: at this point it should be known, this just adds the option on the API auth, it was already there everywhere else.

**Comment by sean-m-sullivan** [comment]
Going back to Alan's comments, OAUTH tokens do work. against AAP, and that is where I want the change to be, on ansible.controller, and maybe it needs to be moved there. hard to do PR's against invisible things.

File: awx_collection/plugins/module_utils/controller_api.py [hunk 4]
```diff
@@ -50,12 +50,41 @@ class ItemNotDefined(Exception):
...
+        request_timeout=dict(
+            type='float',
+            aliases=['aap_request_timeout'],
+            required=False,
+            fallback=(env_fallback, ['CONTROLLER_REQUEST_TIMEOUT', 'AAP_REQUEST_TIMEOUT'])),
+        controller_oauthtoken=dict(
+            type='raw',
+            no_log=True,
+            aliases=['tower_oauthtoken', 'aap_token'],
+             required=False,
```
Comment (djdanielsson) [inline awx_collection/plugins/module_utils/controller_api.py:78]: I think there is an extra space here and the line below it

**Comment by AlanCoding** [comment]
> Going back to Alan's comments, OAUTH tokens do work. against AAP, and that is where I want the change to be, on ansible.controller, and maybe it needs to be moved there. hard to do PR's against invisible things.

Sorry, I shouldn't leave this conversation hanging.

Yes, it is expected that OAuth2 tokens for AAP work. These replace our "direct API" access OAuth2 tokens (targeting the AWX / controller host) which have already been removed in the `devel` branch, and the collection here.

ansible.controller is rolled from the same source. There is no ansible.controller repo to contribute to, because that is this source code. You are in the right place.

Because of this, the remaining change I still want is to purge any reference to "controller_token" because this misleads. There is no controller token, only an AAP token. We can revisit that point for downstream, stable, branches which still have support for the direct API token, and can come out in an ansible.controller release. That is a matter of branch management.

File: awx_collection/plugins/doc_fragments/auth.py (hunk 2 above)
Comment (AlanCoding) [inline awx_collection/plugins/doc_fragments/auth.py]:
```suggestion
```

File: awx_collection/plugins/doc_fragments/auth_plugin.py [hunk 5]
```diff
@@ -17,45 +17,51 @@ class ModuleDocFragment(object):
...
     env:
     - name: CONTROLLER_PASSWORD
-    - name: TOWER_PASSWORD
       deprecated:
         collection_name: 'awx.awx'
         version: '4.0.0'
         why: Collection name change
-        alternatives: 'CONTROLLER_PASSWORD'
+        alternatives: 'TOWER_PASSWORD, AAP_PASSWORD'
+  oauth_token:
```
Comment (AlanCoding) [inline awx_collection/plugins/doc_fragments/auth_plugin.py]:
```suggestion
  aap_token:
```

File: awx_collection/plugins/doc_fragments/auth_plugin.py [hunk 6]
```diff
@@ -17,45 +17,51 @@ class ModuleDocFragment(object):
...
+    description:
+    - The OAuth token to use.
+    env:
+    - name: CONTROLLER_OAUTH_TOKEN
```
Comment (AlanCoding) [inline awx_collection/plugins/doc_fragments/auth_plugin.py]:
```suggestion
    - name: AAP_TOKEN
```

File: awx_collection/plugins/doc_fragments/auth_plugin.py (hunk 3 above)
Comment (AlanCoding) [inline awx_collection/plugins/doc_fragments/auth_plugin.py]:
```suggestion
```

File: awx_collection/plugins/module_utils/controller_api.py [hunk 7]
```diff
@@ -50,12 +50,41 @@ class ItemNotDefined(Exception):
...
+            type='bool',
+            aliases=['tower_verify_ssl', 'aap_validate_certs'],
+            required=False,
+            fallback=(env_fallback, ['CONTROLLER_VERIFY_SSL', 'TOWER_VERIFY_SSL', 'AAP_VALIDATE_CERTS'])),
+        request_timeout=dict(
+            type='float',
+            aliases=['aap_request_timeout'],
+            required=False,
+            fallback=(env_fallback, ['CONTROLLER_REQUEST_TIMEOUT', 'AAP_REQUEST_TIMEOUT'])),
+        controller_oauthtoken=dict(
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/controller_api.py]:
```suggestion
        aap_token=dict(
```

File: awx_collection/plugins/module_utils/controller_api.py [hunk 8]
```diff
@@ -50,12 +50,41 @@ class ItemNotDefined(Exception):
...
+            type='raw',
+            no_log=True,
+            aliases=['tower_oauthtoken', 'aap_token'],
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/controller_api.py]:
```suggestion
```

**Review by AlanCoding — CHANGES_REQUESTED** [review]

File: awx_collection/plugins/module_utils/controller_api.py (hunk 4 above)
Comment (sean-m-sullivan) [reply awx_collection/plugins/module_utils/controller_api.py:78]: not found

**Comment by sean-m-sullivan** [comment]
@AlanCoding Requested changes made.

**Comment by pb82** [comment]
@AlanCoding this looks good to merge now, would you agree? Tests are passing.

**Review by djyasin — APPROVED** [review]

**Comment by djyasin** [comment]
This content was merged in another PR here, so we are going to close this issue.
